### PR TITLE
Update kfp component files to include optional parameter

### PR DESCRIPTION
**Pull request by akchinSTC**
- Update the kfp component definitions to align with best practices
and include descriptions and source license information
- Set directory as source when reading local custom components
- Remove runtime type from reader abstract method

### How was this pull request tested?

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/pipeline/resources/kfp/kubeflow_serve_model_using_kfserving/component.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: Kubeflow - Serve Model using KFServing
```
Comment (ptitzler) [inline elyra/pipeline/resources/kfp/kubeflow_serve_model_using_kfserving/component.yaml:1]:
```suggestion
name: Kubeflow - Serve model using KFServing
```

File: elyra/pipeline/resources/kfp/filter_text_using_shell_and_grep/component.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+name: Filter text using shell and grep
```
Comment (ptitzler) [inline elyra/pipeline/resources/kfp/filter_text_using_shell_and_grep/component.yaml:1]:
```suggestion
name: Filter text 
description: Filter text using shell and grep
```

File: elyra/pipeline/registry.py [hunk 3]
```diff
@@ -793,8 +793,8 @@ def _get_reader(self, component):
         Find the proper reader based on the given registry component.
         """
         try:
-            component_type = list(component['path'].keys())[0]
-            return self.readers.get(component_type)
+            reader_type = list(component['path'].keys())[0]
```
Comment (ptitzler) [inline elyra/pipeline/registry.py:796]: Since the change is not related to the PR title it should be listed in the PR description

**Comment by ptitzler** [comment]
Could we add a comment at the beginning of each example component file:
- Since Elyra re-uses existing components that were published by somebody else, we should document the raw URL where the YAML file originally came from , e.g. 
  ```
  # Component source location: https://...
  ```
- Often there is not enough information in the specification for the user to make an informed decision about valid values. If one exists, we should  reference a web page (if applicable) that contains more details. For example, https://github.com/kubeflow/pipelines/blob/master/components/kubeflow/kfserving/README.md
   ```
   # Component details: https://
   ```

File: etc/config/components/kfp_component_catalog.json [hunk 4]
```diff
@@ -3,19 +3,19 @@
     "run-notebook-using-papermill": {
       "name": "Run Notebook Using Papermill",
       "path": {
-        "url": "https://raw.githubusercontent.com/kubeflow/pipelines/1.4.1/components/notebooks/Run_notebook_using_papermill/component.yaml"
+        "filename": "resources/kfp/run_notebook_using_papermill/component.yaml"
```
Comment (kevin-bates) [inline etc/config/components/kfp_component_catalog.json:6]: I would suggest we move the `resources/kfp` portion of these into the server code that, today, uses `__file__` in #1801.  That way, _when_ we move these files out of our installed location, we only need to update the server code that concatenates to `__file__` and don't have to change both the server and the catalog file entries themselves.

**Comment by lresende** [comment]
@akchinSTC Could you please update the license and list the origin + license of these component files

File: elyra/pipeline/resources/kfp/filter_text_using_shell_and_grep/component.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Component source location: https://raw.githubusercontent.com/kubeflow/pipelines/master/components/sample/Shell_script/component.yaml
+name: Filter text
```
Comment (ptitzler) [inline elyra/pipeline/resources/kfp/filter_text_using_shell_and_grep/component.yaml:2]: Please add a `description` (see earlier comment)

File: etc/config/components/kfp_component_catalog.json (hunk 4 above)
Comment (ptitzler) [reply etc/config/components/kfp_component_catalog.json:6]: I agree

File: elyra/pipeline/registry.py [hunk 6]
```diff
@@ -768,7 +772,10 @@ def get_properties(self, processor_type, component_id):
 
             component_path = component['path'][reader._type]
             if reader._type == "filename":
-                component_path = os.path.join(os.path.dirname(__file__), component_path)
+                component_resource_dir = ""
+                if processor_type in ['kfp', 'airflow']:
+                    component_resource_dir = "resources/" + processor_type
```
Comment (kevin-bates) [inline elyra/pipeline/registry.py:777]: I'm confused between `processor_type` and `parser._type` and would prefer that we consolidate these if possible.  Since 'local' is not a formal processor type, I suspect it's an invariant condition that this method is ever called when `processor_type == 'local'` - is that a correct statement?

We should also try to avoid the `in` clause, otherwise we'll require code changes on each new processor.

This question seems like one fundamental to the overall design, so I'm not sure it can be addressed here.

Also, I believe this area of the code is being changed in #1801 (and eliminated), so it may be moot altogether.

Comment (kiersten-stokes) [reply elyra/pipeline/registry.py:777]: Yeah, you're correct that the way we implement this particular portion of the code will be changed once the refactoring is done. It will remove the need to use `in` and we'll put the onus on the component parser to determine it's own path to the local file, then just pass that path to the reader

Comment (akchinSTC) [reply elyra/pipeline/registry.py:777]: will wait for the refactoring to complete then to make the changes, thx

**Comment by lresende** [comment]
@akchinSTC @kiersten-stokes Could you please resolve conflicts.

File: elyra/pipeline/component_parser_kfp.py [hunk 7]
```diff
@@ -13,15 +13,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-import os
-from typing import List
-
 import yaml
+from typing import List
```
Comment (kiersten-stokes) [inline elyra/pipeline/component_parser_kfp.py:17]: Per the linter's suggestion

```suggestion

from typing import List
```

**Review by kiersten-stokes — CHANGES_REQUESTED** [review]
I think the linting issue below is the only outstanding thing for this PR (I say as if I didn't introduce the problem...)

File: elyra/pipeline/component.py [hunk 8]
```diff
@@ -242,7 +242,7 @@ def type(self) -> str:
         return self.type
 
     @abstractmethod
-    def read_component_definition(self, component_id: str, location: str) -> str:
+    def read_component_definition(self, component_id: str, location: str, runtime_type: str) -> str:
```
Comment (lresende) [inline elyra/pipeline/component.py:245]: The reader is agnostic of the runtime, so this parameter should not be added.

**Review by lresende — CHANGES_REQUESTED** [review]
We need to make the registry aware of the additional location where components are located either by convention and/or by explicitly adding things on the registry JSON with the proper type and not let a reader calculate the location based on the type of the runtime, because then that reader looses its flexibility

File: elyra/pipeline/component.py (hunk 8 above)
Comment (kiersten-stokes) [reply elyra/pipeline/component.py:245]: Woops, looks like I kept this in accidentally. Pushing fix now

**Review by kiersten-stokes — APPROVED** [review]
LGTM! Tested an example for each runtime and each succeeded

**Review by lresende — APPROVED** [review]
I am not sure if for file-based directory of components we need to list one by one in the registry, but for this PR I guess it's ok to avoid delays, but we will need to revisit this in the future.
